core/templates/pages/exploration-player-page/services/exploration-engine.service.spec.ts

```diff
@@ -0,0 +1,909 @@
+
+  it('should load exploration when initialized in ' +
+    'exploration editor page', () => {
+    let initSuccessCb = jasmine.createSpy('success');
+    let paramChanges = paramChangeObjectFactory.createFromBackendDict(
+      paramChangeDict);
+    // Setting exploration editor page.
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(true);
+    spyOn(urlService, 'getPathname')
+      .and.returnValue('/create/in/path/name');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(false);
+
+    // Since the constructor will be automatically called in unit tests, it
+    // is hard to test or spy on the constructor. So, we have created a
+    // function to manually trigger and tests different edge cases.
+    explorationEngineService.setExplorationProperties();
+
+    expect(explorationEngineService.isInPreviewMode()).toBe(true);
+    expect(() => {
+      explorationEngineService.getExplorationTitle();
+    }).toThrowError('Cannot read property \'title\' of undefined');
+
+    explorationEngineService.initSettingsFromEditor('Start', [paramChanges]);
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+    const explorationTitle = explorationEngineService.getExplorationTitle();
+    expect(explorationTitle).toBe('My Exploration Title');
+    expect(initSuccessCb).toHaveBeenCalled();
+  });
+
+  describe('on submitting answer ', () => {
+    it('should call success callback if the submitted ' +
+      'answer is correct', () => {
+      let initSuccessCb = jasmine.createSpy('success');
+      let submitAnswerSuccessCb = jasmine.createSpy('success');
+      let answer = 'answer';
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory.createFromBackendDict({
+          dest: 'Mid',
+          feedback: {
+            content_id: 'feedback_1',
+            html: 'Answer is correct!'
+          },
+          labelled_as_correct: true,
+          param_changes: [],
+          refresher_exploration_id: null,
+          missing_prerequisite_skill_id: null
+        }), 1, 0, 'default_outcome');
+
+      let lastCard = stateCardObjectFactory.createNewCard(
+        'Card 1', 'Content html', 'Interaction text', null,
+        null, null, 'content_id');
+
+      spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+      spyOn(playerTranscriptService, 'getLastStateName')
+        .and.returnValue('Start');
+      spyOn(playerTranscriptService, 'getLastCard').and.returnValue(lastCard);
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+
+      explorationEngineService.init(
+        explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+      const isAnswerCorrect = explorationEngineService.submitAnswer(
+        answer, textInputService, submitAnswerSuccessCb);
+
+      expect(submitAnswerSuccessCb).toHaveBeenCalled();
```

Comment:
Is it possible to test **all** the callbacks with `.toHaveBeenCalledWith`?

Reply:
Most of the tests do have `.toHaveBeenCalledWith`.
In the specific test case, success callbacks consist of HTML tags. These have random spacing which makes them hard to test.
![Screenshot from 2021-06-17 04-16-42](https://user-images.githubusercontent.com/54100702/122304887-95dc9b80-cf23-11eb-8771-866fc42860b4.png)
And I think checking implementation details of success callback is not really necessary as we are only testing whether the submitted answer is correct or not.